juju/model.py

```diff
@@ -0,0 +1,167 @@
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(Mapping):
+    def __init__(self, relation_name, relation_id, backend):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._keys = self._backend.get_relation_members(self.relation_id)
+
+    def keys(self):
+        return self._keys
+
+    def __contains__(self, key):
+        return key in self._keys
+
+    def __iter__(self):
+        return iter(self._keys)
+
+    def __len__(self):
+        return len(self._keys)
+
+    def __getitem__(self, member):
+        if member not in self._keys:
+            raise KeyError(member)
+        return self._backend.relation_get(self.relation_id, member.name)
+
+
+class ModelError(Exception):
+    pass
+
+
+class TooManyRelatedApps(ModelError):
+    def __init__(self, relation_name, num_related, max_supported):
+        super().__init__(f'Too many remote applications on {relation_name} ({num_related} > {max_supported})')
+        self.relation_name = relation_name
+        self.num_related = num_related
+        self.max_supported = max_supported
+
+
+class ModelBackend:
+    def __init__(self, local_unit_name):
+        local_app_name = local_unit_name.split('/')[0]
+        self.local_app = Application(local_app_name)
+        self.local_unit = Unit(local_unit_name, self.local_app)
+        self._apps = {local_app_name: self.local_app}
+        self._units = {local_unit_name: self.local_unit}
+
+    def relation_ids(self, relation_name):
+        output = run(['relation-ids', relation_name, '--format=json'], stdout=PIPE, check=True).stdout
+        return json.loads(output.decode('utf8'))
+
+    def relation_list(self, relation_id):
+        output = run(['relation-list', '-r', relation_id, '--format=json'], stdout=PIPE, check=True).stdout
+        return json.loads(output.decode('utf8'))
+
+    @lru_cache(maxsize=None)
+    def relation_get(self, relation_id, member_name):
+        output = run(['relation-get', '-r', relation_id, '-', member_name, '--format=json'], stdout=PIPE, check=True).stdout
+        return json.loads(output.decode('utf8'))
+
+    @lru_cache(maxsize=None)
+    def get_relations(self, relation_name):
+        relation_ids = self.relation_ids(relation_name)
+        return [Relation(relation_name, relation_id, self) for relation_id in relation_ids]
+
+    @lru_cache(maxsize=None)
+    def get_relation_members(self, relation_id):
+        # The local unit is always a member of this charm's relations.
+        # TODO: Juju will add support for application-level relation data, at
+        # which point the local app will also become a member of all relations.
+        members = {self.local_unit}
+        for remote_unit_name in self.relation_list(relation_id):
```

Comment:
Same.. this method feels like entirely a frontend thing. Note that all of its functionality is performed on top of that obtained from a different public method on the backend. That means this same method will do the same job independent of which backend implements the relation_list method.